File: backend/app/Models/Voucher.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Voucher extends Model
{
    use HasFactory;

    protected $fillable = [
        'code',
        'name',
        'type',
        'value',
        'min_transaction',
        'max_discount',
        'usage_limit',
        'usage_count',
        'user_limit',
        'product_id',
        'customer_type',
        'is_active',
        'start_date',
        'end_date',
        'description',
    ];

    protected $casts = [
        'value' => 'decimal:2',
        'min_transaction' => 'decimal:2',
        'max_discount' => 'decimal:2',
        'usage_limit' => 'integer',
        'usage_count' => 'integer',
        'user_limit' => 'integer',
        'product_id' => 'integer',
        'is_active' => 'boolean',
        'start_date' => 'datetime',
        'end_date' => 'datetime',
    ];

    // ==================== RELATIONSHIPS ====================

    /**
     * Get all voucher usages
     */
    public function usages()
    {
        return $this->hasMany(VoucherUsage::class);
    }

    /**
     * Get the associated product (if any)
     */
    public function product()
    {
        return $this->belongsTo(Product::class);
    }

    // ==================== SCOPES ====================

    /**
     * Scope a query to only include active vouchers
     */
    public function scopeActive($query)
    {
        return $query->where('is_active', true)
            ->where(function ($q) {
                $q->whereNull('start_date')
                    ->orWhere('start_date', '<=', now());
            })
            ->where(function ($q) {
                $q->whereNull('end_date')
                    ->orWhere('end_date', '>=', now());
            })
            ->where(function ($q) {
                $q->whereNull('usage_limit')
                    ->orWhereRaw('usage_count < usage_limit');
            });
    }

    // ==================== METHODS ====================

    /**
     * Check if voucher is valid
     */
    public function isValid()
    {
        return $this->is_active
            && (is_null($this->start_date) || $this->start_date <= now())
            && (is_null($this->end_date) || $this->end_date >= now())
            && (is_null($this->usage_limit) || $this->usage_count < $this->usage_limit);
    }

    /**
     * Calculate discount amount
     */
    public function calculateDiscount($transactionAmount)
    {
        if ($transactionAmount < $this->min_transaction) {
            return 0;
        }

        if ($this->type === 'percentage') {
            $discount = ($transactionAmount * $this->value) / 100;

            return $this->max_discount ? min($discount, $this->max_discount) : $discount;
        }

        // Fixed amount
        return $this->value;
    }

    /**
     * Increment usage count
     */
    public function incrementUsage()
    {
        $this->increment('usage_count');
    }
}

File: enstore/lib/core/models/voucher.dart
class VoucherModel {
  final int id;
  final String code;
  final String name;
  final String type;
  final double value;
  final String? description;
  final DateTime? startDate;
  final DateTime? endDate;
  final double minTransaction;
  final double? maxDiscount;
  final bool isAvailable;
  final int userRemaining;

  VoucherModel({
    required this.id,
    required this.code,
    required this.name,
    required this.type,
    required this.value,
    this.description,
    this.startDate,
    this.endDate,
    required this.minTransaction,
    this.maxDiscount,
    this.isAvailable = true,
    this.userRemaining = 0,
  });

  factory VoucherModel.fromJson(Map<String, dynamic> json) {
    return VoucherModel(
      id: json['id'],
      code: json['code'],
      name: json['name'],
      type: json['type'],
      value: double.tryParse(json['value'].toString()) ?? 0.0,
      description: json['description'],
      startDate: json['start_date'] != null ? DateTime.parse(json['start_date']) : null,
      endDate: json['end_date'] != null ? DateTime.parse(json['end_date']) : null,
      minTransaction: double.tryParse(json['min_transaction'].toString()) ?? 0.0,
      maxDiscount: json['max_discount'] != null ? double.tryParse(json['max_discount'].toString()) : null,
      isAvailable: json['is_available'] ?? true,
      userRemaining: json['user_remaining'] ?? 0,
    );
  }
}
